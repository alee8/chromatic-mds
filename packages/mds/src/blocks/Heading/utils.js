import { EMPTY_STRING } from '~mds/constants';
/**
 * Remove all occurrences of Htags.
 * @param {String} str HTML markup that includes h[1-6] tag
 * @returns {String} original string with all h[1-6] tags removed
 */
export const removeHTag = (str) => {
  return str.replace(/<h[1-6]>|<h[1-6] .*?>|<\/h[1-6]>/gm, EMPTY_STRING);
};
