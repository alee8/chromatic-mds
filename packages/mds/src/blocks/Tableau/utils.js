/**
 * Tableau desktop slug.
 *
 * @constant
 */
const DESKTOP_VIZ = 'Desktop';
/**
 * Tableau mobile slug.
 *
 * @constant
 */
const MOBILE_VIZ = 'Mobile';

/**
 * The constant is internal to Tableau implementation.
 * It references the currently applicable Tableau API url.
 *
 * TABLEAU_API is embedded in the official Tableau API https://public.tableau.com/javascripts/api/tableau-2.min.js.
 * The public Tableau API cannot be used as it does a document.write; React app-friendly action.
 * NOTE: Tableau periodically updates tableau-2.min.js without any official announcement (specifically the value used by TABLEAU_API).
 * Therefore, when TABLEU_API does not load, it is time to view the internals of tableau-2.min.js script for changes.
 *
 * @constant
 */
export const TABLEAU_API = 'https://public.tableau.com/javascripts/api/tableau-2.9.1.min.js';

/**
 * Prefix embed url to Desktop and Mobile Tableau dashboard.
 * The constant is internal to Tableau implementation.
 *
 * @constant
 */
const TABLEAU_URL = 'https://public.tableau.com/views';

/**
 * Instantiate each Tableau Viz object with the provided parameters.
 *
 * @param {Object} config Config parameters for instantiating Tableau Viz.
 * @returns
 */
export function instantiateViz({
  desktop: {
    ref: desktopRef,
    size: {
      height: desktopHeight,
      width: desktopWidth,
    },
  },
  embedShortname,
  mobile: {
    ref: mobileRef,
    size: {
      height: mobileHeight,
      width: mobileWidth,
    },
  },
}) {
  const { tableau = undefined } = window;
  if (!tableau) {
    // window.tableau not available; return and wait to create script tag to get tableau API or if created, for scriptLoaded state to change
    return false;
  }

  const urlPrefix = `${TABLEAU_URL}/${embedShortname}`;

  // Instantiate a Viz instance for desktop and mobile Tableau Dashbooards.
  /* eslint-disable no-new */
  new tableau.Viz(
    desktopRef.current,
    `${urlPrefix}/${DESKTOP_VIZ}`,
    {
      height: desktopHeight,
      width: desktopWidth,
    }
  );
  new tableau.Viz(
    mobileRef.current,
    `${urlPrefix}/${MOBILE_VIZ}`,
    {
      height: mobileHeight,
      width: mobileWidth,
    }
  );
  /* eslint-enable no-new */
  return true;
}
