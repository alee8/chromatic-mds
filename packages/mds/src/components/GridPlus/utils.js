import { GRID_SIZE_OPTIONS } from '~mds/constants/sizes';

export const validGridSizes = (size) => GRID_SIZE_OPTIONS.includes(size);
