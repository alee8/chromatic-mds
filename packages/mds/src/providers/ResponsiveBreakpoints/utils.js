/**
 */
import {
  WRENCH_BREAKPOINT_SETTINGS,
  WRENCH_BREAKPOINT_VALUES,
} from '~mds/constants/wrench-breakpoints';

/** ----------------- Private ResponsiveBreakpointsProvider utilities -------------------------
 *
 * These utils are exported to support ResponsiveBreakpoints.
 * DO NOT IMPORT this file directly.
*/

/**
 * Small helper to formulate the CSS min-width media query given a breakpoint.
 *
 * @param   {number} breakpoint
 * @returns {string} media query
 */
function getMinWidthMediaQuery(breakpoint) {
  return `(min-width: ${breakpoint}px)`;
}

export function getMinWidthMediaQueryObject() {
  return {
    [WRENCH_BREAKPOINT_VALUES.xs]: getMinWidthMediaQuery(WRENCH_BREAKPOINT_SETTINGS.xs),
    [WRENCH_BREAKPOINT_VALUES.sm]: getMinWidthMediaQuery(WRENCH_BREAKPOINT_SETTINGS.sm),
    [WRENCH_BREAKPOINT_VALUES.md]: getMinWidthMediaQuery(WRENCH_BREAKPOINT_SETTINGS.md),
    [WRENCH_BREAKPOINT_VALUES.lg]: getMinWidthMediaQuery(WRENCH_BREAKPOINT_SETTINGS.lg),
    [WRENCH_BREAKPOINT_VALUES.xl]: getMinWidthMediaQuery(WRENCH_BREAKPOINT_SETTINGS.xl),
  };
}

/**
 * Handler that is called when media-query match results change.
 * Each media query is checked against the document using the parsed media query's `matches` method for its match result.
 * The resulting `updatedMatches` is recorded using the `setMediaQueryMatches`.
 *
 * @param {string}   bpNameKeys from custom hook's provided `mediaQueries`
 * @param {object}   parsedMediaQueryList set of parsed `mediaQueries`
 * @param {function} setMediaQueryMatches records `updatedMatches`
 */
export function handleMediaQueryListener(bpNameKeys, parsedMediaQueryList, setMediaQueryMatches) {
  const updatedMatches = bpNameKeys.reduce((acc, media) => {
    acc[media] = !!(parsedMediaQueryList?.[media]?.matches);
    return acc;
  }, {});
  setMediaQueryMatches(updatedMatches);
}
