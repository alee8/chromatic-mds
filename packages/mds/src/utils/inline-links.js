export function parseHtml(html, options) {
  return html;
}
