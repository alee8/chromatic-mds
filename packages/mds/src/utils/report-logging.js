export function debugLog(message, data = EMPTY_STRING) {
  const timestamp = timeFormat(Date.now());

  /* eslint-disable-next-line no-console */
  console.log(`ðŸ”µ  ${timestamp}  ${message}`, data);
}

export function reportError({
  error,
  severity = 'error',
}) {
  console.error(`Error ${error} with severity of ${severity}`);
}
