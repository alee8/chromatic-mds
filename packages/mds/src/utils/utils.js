import {
  EMPTY_STRING,
  UNIT_KEY,
} from '~mds/constants';

/**
 * Maps hyphenated strings to camel case.
 * If char following hyphen is letter, it is upperCased.
 * If char following hyphen is number, it is left as number.
 * @param {String} str
 * @returns camelCasedString
 */
export function camelCase(str) {
  return str.replace(
    /-([a-z\d])/gi,
    (match, matchedChar) => (Number.isInteger(matchedChar) ? matchedChar : matchedChar.toUpperCase())
  );
}

/**
 * Takes an arr or object, produces an object using array's values or object's key to produce object whose (key, value) pair has same key for the pair
 * @param {Array} arrOrObj
 * @returns {Object} result
 */
export const generateObject = (arrOrObj) => {
  const paramType = typeof arrOrObj;
  if (!arrOrObj || paramType !== 'object') {
    return {};
  }
  if (Array.isArray(arrOrObj)) {
    return arrOrObj.reduce((result, key) => {
      // Following to get around airbnb `no-param-assign: {props: true}`
      // https://github.com/airbnb/javascript/issues/719#issuecomment-237535701
      if (key !== UNIT_KEY) {
        const keyType = typeof key;
        if (key && key.length > 0) {
          // with value for key
          Object.assign(result, { [camelCase(key)]: key });
        } else if (keyType === 'string' || keyType === 'undefined') {
          // with EMPTY_STRING or undefined for key
          Object.assign(result, { default: EMPTY_STRING });
        } else if (typeof key === 'object') {
          // with null or {} or [] for key
          Object.assign(result, { none: EMPTY_STRING });
        } else {
          // catch all case; set key to error and value of keyType to flag
          Object.assign(result, { error: keyType });
        }
      }
      return result;
    }, {});
  }
  return Object.keys(arrOrObj).reduce((result, key) => {
    // Following to get around airbnb `no-param-assign: {props: true}`
    // https://github.com/airbnb/javascript/issues/719#issuecomment-237535701
    Object.assign(result, { [camelCase(key)]: key });
    return result;
  }, {});
};

/**
 * Place hyphen before the number. String prefix is all lowercase.
 * Transforms:
 *   'size9' -> 'size-9'
 *   'size10' -> 'size-10'
 *   'size' -> 'size'
 *   'Size' -> 'Size'
 *
 * @param {String} str
 * @returns {String} result
 */
export const hyphenateNumber = (str) => {
  const re = /^([a-z]+)(\d+)$/;
  return str.replace(re, '$1-$2');
};

/**
 * Remove hyphen before the number.
 * Transforms:
 *   'size-9' -> 'size9'
 *   'size-10' -> 'size10'
 *   'si-10ze' -> 'si-10ze'
 *   'size' -> 'size'
 *   'Size' -> 'Size'
 *
 * @param {String} str
 * @return {String} result
 */
export const unHyphenateNumber = (str) => {
  const re = /^([a-z]+)-(\d+)$/;
  return str.replace(re, '$1$2');
};
