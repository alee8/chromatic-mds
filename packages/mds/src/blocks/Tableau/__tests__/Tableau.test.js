/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';

import '~tests/setupTests/matchMedia';
import ComponentWrappers from '~tests/setupTests/ComponentWrappers';
import createScriptTag from '~mds/providers/LoadScript/create-script-tag';

import LoadScriptProvider, { useLoadScriptContext } from '~mds/providers/LoadScript';

import {
  BasicExample,
  ErrorExample,
} from './Examples';

import { TABLEAU_API } from '../utils';

jest.mock('~mds/providers/LoadScript/create-script-tag');
beforeEach(() => {
  jest.useFakeTimers();
});
afterEach(() => {
  jest.useRealTimers();
});

describe('Render Tableau sanity checks', () => {
  test('Basic Tableau', () => {
    const { container, unmount } = render(
      <ComponentWrappers>
        <BasicExample />
      </ComponentWrappers>
    );

    expect(container).toMatchSnapshot();
    unmount(container);
  });

  test('Tableau Error', () => {
    const { container, unmount } = render(
      <ErrorExample isDesktop />
    );

    expect(container).toMatchSnapshot();
    unmount(container);
  });
});

describe('LoadScript Tableau', () => {
  test('Render with TABLEAU_API load', () => {
    const { container } = render(
      <LoadScriptProvider>
        <TestComponentScriptTag />
      </LoadScriptProvider>
    );
    expect(container.textContent).toStrictEqual('{"tableauScriptRef":{"current":null},"wistiaScriptRef":{"current":null}}');
  });

  test('createScriptTag TABLEAU_API', async () => {
    const onload = jest.fn();
    const onerror = jest.fn();
    createScriptTag(TABLEAU_API, onload, onerror);
    jest.runAllTimers();
    expect(onload).toBeCalledTimes(1);
    expect(onerror).toBeCalledTimes(0);
  });
});

function TestComponentScriptTag() {
  const ref = useLoadScriptContext();
  return JSON.stringify(ref);
}
