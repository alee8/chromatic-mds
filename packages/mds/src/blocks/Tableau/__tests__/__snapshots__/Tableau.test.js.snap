// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Tableau sanity checks Basic Tableau 1`] = `
<div>
  <div
    class="mntv-mdsGridPlusContainer-0-2-14 mntv-mdsGridPlusContainer-d5-0-2-16"
  >
    <div
      class="mntv-mdsGridPlus-0-2-15 mntv-mdsGridPlus-d11-0-2-17 mntv-addFlex-0-2-12 mntv-addFlex-d0-0-2-13"
    >
      <div
        class="mntv-mdsGridPlusItem-0-2-20 mntv-mdsGridPlusItem-d0-0-2-21 mntv-addFlex-0-2-18 mntv-addFlex-d0-0-2-19"
      >
        <div
          class="mntv-contentWrapper-0-2-1"
        >
          <div
            class="mntv-tableauHide-0-2-5 mntv-tableauContainer-0-2-4 mntv-desktopSize-0-2-2 mntv-desktopSize-d0-0-2-10 mntv-tableauDesktop-0-2-8"
          />
          <div
            class="mntv-tableauShow-0-2-6 mntv-tableauContainer-0-2-4 mntv-nonDesktopSize-0-2-3 mntv-nonDesktopSize-d1-0-2-11 mntv-tableauNonDesktop-0-2-9"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render Tableau sanity checks Tableau Error 1`] = `
<div>
  <div
    class="mntv-errorContainer-0-2-7 mntv-desktopSize-0-2-2 mntv-desktopSize-d0-0-2-10 mntv-wrenchBox-0-2-14 mntv-wrenchBox-d0-0-2-15 mntv-wrenchSpacing-0-2-12 mntv-wrenchSpacing-d0-0-2-13"
  >
    <div
      class="mntv-wrenchBox-0-2-14 mntv-wrenchBox-d1-0-2-17 mntv-wrenchSpacing-0-2-12 mntv-wrenchSpacing-d1-0-2-16"
    >
      <div
        class="mntv-wrenchBox-0-2-14 mntv-wrenchBox-d2-0-2-19 mntv-wrenchSpacing-0-2-12 mntv-wrenchSpacing-d2-0-2-18"
      >
        <p
          class="mntv-background-0-2-20 mntv-background-d1-0-2-24  mntv-dropCap-0-2-21 mntv-dropCap-d4-0-2-25  mntv-paragraph-0-2-22 mntv-paragraph-d5-0-2-26  mntv-paragraphTypography-0-2-23 mntv-paragraphTypography-d6-0-2-27  "
        >
          ERROR
        </p>
      </div>
      <div
        class="noneMarginBottom noneMarginTop nonePaddingBottom nonePaddingTop mntv-heading-0-2-28 mntv-heading-d0-0-2-31  mntv-headingSize-0-2-29 mntv-headingSize-d1-0-2-32  "
        id=""
      >
        <h3>
          Sorry, we cannot find that data visualization.
        </h3>
      </div>
      <div
        class="mntv-wrenchBox-0-2-14 mntv-wrenchBox-d3-0-2-34 mntv-wrenchSpacing-0-2-12 mntv-wrenchSpacing-d3-0-2-33"
      >
        <p
          class="mntv-background-0-2-20 mntv-background-d9-0-2-35  mntv-dropCap-0-2-21 mntv-dropCap-d12-0-2-36  mntv-paragraph-0-2-22 mntv-paragraph-d13-0-2-37  mntv-paragraphTypography-0-2-23 mntv-paragraphTypography-d14-0-2-38  "
        >
          Please refresh your browser or check back later to see if the situation is resolved.
        </p>
      </div>
    </div>
  </div>
</div>
`;
